File: java/sum-of-multiples/src/main/java/SumOfMultiples.java
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class SumOfMultiples {

    private int sum;

    SumOfMultiples(int number, int[] set) {
        this.sum = Arrays.stream(set)
                    .mapToObj(n -> getMultiplesUtilLimit(n, number))
                    .flatMap(list -> list.stream())
                    .mapToInt(i -> i).distinct().sum();
    }

    int getSum() {
        return sum;
    }

    private List<Integer> getMultiplesUtilLimit(int number, int limit) {
        return number == 0 
                ? List.of()
                : IntStream.range(1, limit)
                    .filter(n -> n % number == 0).boxed().toList();
    }
}

File: c/sum-of-multiples/sum_of_multiples.c
#include "sum_of_multiples.h"

#include <stdbool.h>

unsigned int sum(const unsigned int *factors, const size_t number_of_factors,
                 const unsigned int limit) {

    unsigned int sum = 0;

    for (unsigned int i = 1; i < limit; i++) {
        for (size_t j = 0; j < number_of_factors; j++) {
            unsigned int factor = factors[j];

            if (factor != 0 && i % factor == 0) {
                sum += i;
                break;
            }

        }

    }

    return sum;
}

